lib/pos_fee_estimator.go

```diff
@@ -171,29 +183,254 @@ func (posFeeEstimator *PoSFeeEstimator) sortCachedBlocks() {
+func computeFeeGivenTxnAndFeeRate(txn *MsgDeSoTxn, feeRateNanosPerKB uint64) (uint64, error) {
+	// Create a clone of the txn, so we don't modify the original.
+	txnClone, err := txn.Copy()
+	if err != nil {
+		return 0, errors.Wrap(err, "computeFeeGivenTxnAndFeeRate: Problem copying txn")
+	}
+	// Set the nonce to the maximum value if it's not already set.
+	if txnClone.TxnNonce == nil {
+		txnClone.TxnNonce = &DeSoNonce{
+			ExpirationBlockHeight: math.MaxUint64,
+			PartialID:             math.MaxUint64,
+		}
+	}
+	if txnClone.TxnNonce.ExpirationBlockHeight == 0 {
+		txnClone.TxnNonce.ExpirationBlockHeight = math.MaxUint64
+	}
+	if txnClone.TxnNonce.PartialID == 0 {
+		txnClone.TxnNonce.PartialID = math.MaxUint64
+	}
+
+	// Set the TxnFeeNanos to the maximum value.
+	txnClone.TxnFeeNanos = math.MaxUint64
+	txnFeeNanos, err := computeFeeRecursive(txnClone, feeRateNanosPerKB)
+	if err != nil {
+		return 0, errors.Wrap(err, "computeFeeGivenTxnAndFeeRate: Problem computing fee rate recursively")
+	}
+	return txnFeeNanos, nil
+}
+
+// computeFeeRecursive computes the fee in nanos for the provided transaction and fee rate
+// in nanos per KB. It does this by recursively computing the fee until the fee converges.
+// It should only be called from computeFeeGivenTxnAndFeeRate and assumes that the TxnFeeNanos
+// field of the txn is set to the maximum value on the first call.
+// It computes the length of the unsigned transaction bytes and then adds the maximum DER sig
+// length to estimate an upper bound on the size of the signed transaction. Next,
+// it will compute the fee by taking the fee rate in nanos per KB and multiplying it by the
+// estimated size of the signed transaction in KB. If the fee computed here is less than the
+// TxnFeeNanos field of the txn, it will set the TxnFeeNanos field to the computed fee and
+// call itself recursively. This will continue until the computed fee is greater than or equal
+// to the TxnFeeNanos field of the txn. At this point, it will return the TxnFeeNanos field.
+// This will converge in 9 iterations or fewer. The maximum number of iterations is 9 because
+// the transaction fee is the only field that gets modified so there are only 8 possible
+// byte lengths that can be used to represent the transaction fee, and we ensure that the
+// number of bytes that are required to represent computed transaction fee field is always less than
+// the number of bytes to represent the TxnFeeNanos field at the previous iteration.
+func computeFeeRecursive(txn *MsgDeSoTxn, feeRateNanosPerKB uint64) (uint64, error) {
+	// Get the length of bytes in the txn.
+	txnBytesNoSignature, err := txn.ToBytes(true)
+	if err != nil {
+		return 0, errors.Wrap(err, "computeFeeRecursive: Problem serializing txn")
+	}
+	const MaxDERSigLen = 74
+	txnBytesLen := uint64(len(txnBytesNoSignature)) + MaxDERSigLen
+
+	txnFee := txnBytesLen * feeRateNanosPerKB / 1000
+
+	if (txnBytesLen*feeRateNanosPerKB)%1000 != 0 {
+		txnFee++
+	}
```

Comment:
Maybe create a const for BytesPerKB to make the math a bit clearer.
```suggestion
	// Compute the new txn fee. If the computed fee is a decimal, we round up to the
	// next integer value. We define the math as follows:
	// - We need to compute CEIL(txnBytesLen * feeRateNanosPerKB / BytesPerKB)
	// - We use integer math to compute FLOOR(txnBytesLen * feeRateNanosPerKB + BytesPerKB) / BytesPerKB)
	txnFeeInBytes := (txnBytesLen * feeRateNanosPerKB + BytesPerKB) / BytesPerKB
```